### PR TITLE
Issue #83

### PR DESCRIPTION
**Pull request by Alaboudi1**

### CODE REVIEW

File: src/jsEditor/js-editor.js [hunk 1]
```diff
@@ -50,33 +51,42 @@ export class JsEditor {
...
-       
-        // subscribe to this event to be notified with the following data when the JS-editor changed.   
-        //TODO: make this smarter by only publishing the event when there is an actual input i.e. not empty space.
-        
-         ea.publish('onJsEditorChanged', {
-            js: js, 
-            length: session.getLength(), 
+        let curs = editor.getCursorPosition().row + 1;
+
+        // This line strip out the spaces at the end of the documents
```
Comment (luminaxster) [inline src/jsEditor/js-editor.js]: remove the comments. On my side, you are good to go =]

File: index.html [hunk 2]
```diff
@@ -1,35 +1,43 @@
...
+	<script src="https://cdn.firebase.com/js/client/2.3.2/firebase.js"></script>
+
+	<!-- ACE and its JavaScript mode and theme files -->
+	<script src="https://cdnjs.cloudflare.com/ajax/libs/ace/1.1.3/ace.js"></script>
+
+	<!-- Firepad -->
+	<script src="https://cdn.firebase.com/libs/firepad/1.2.0/firepad.js"></script>
+
+	<!-- MD5 Hash Algo -->
+	<script src="http://crypto-js.googlecode.com/svn/tags/3.1.2/build/rollups/md5.js"></script>
```
Comment (jbhilton11) [inline index.html]: Might want to put this with jQuery and the other globally-defined libraries.

File: src/jsEditor/js-editor.js [hunk 3]
```diff
@@ -9,29 +9,30 @@ export class JsEditor {
   constructor(eventAggregator) {
     this.eventAggregator = eventAggregator;
     this.hasErrors = false;
+    this.editorText = '1';
```
Comment (jbhilton11) [inline src/jsEditor/js-editor.js]: Why is this.editorText being set to '1'?

File: src/jsEditor/js-editor.js [hunk 4]
```diff
@@ -50,33 +51,42 @@ export class JsEditor {
...
-            js: js, 
-            length: session.getLength(), 
+        let curs = editor.getCursorPosition().row + 1;
+
+        // This line strip out the spaces at the end of the documents
+        let newStr = js.replace(/(\s+$)/g, '');
+        // then, hash it and store it in hash variable 
+        let hash = CryptoJS.MD5(newStr);
+
+        if (editorText > hash || editorText < hash) {
```
Comment (jbhilton11) [inline src/jsEditor/js-editor.js]: This should be editorText !== hash. It's much easier to read and compare one thing than two :).

File: src/jsGutter/js-gutter.js [hunk 5]
```diff
@@ -1,120 +1,152 @@
...
+
+        ea.subscribe('onCursorMoved', info => {
+
+            let lastDiv = this.getLastDiv();
+            let iframeBody = $('#gutter');
+            let line = info.cursor;
+            let lastline = info.lastVisibleRow;
+
+            if (iframeBody.find('#line' + lastline).length == 0) {
+                this.CreateLine(lastline);
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please use camel-case for function names.

File: src/jsGutter/js-gutter.js [hunk 6]
```diff
@@ -1,120 +1,152 @@
...
+            if (lastline < lastDiv) {
+                this.removeLine(lastline, lastDiv);
+            }
+
+            iframeBody.find("#line" + this.selectedLine).removeClass("highlight_gutter");
+            iframeBody.find("#line" + line).addClass("highlight_gutter");
+            this.selectedLine = line;
+
+
+            this.LastVisibleRow = info.lastVisibleRow;
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please use camel-case for variable name.

File: src/jsGutter/js-gutter.js [hunk 7]
```diff
@@ -1,120 +1,152 @@
...
+
+
+
+    subscribe() {
+        let ea = this.eventAggregator;
+
+        ea.subscribe('onCursorMoved', info => {
+
+            let lastDiv = this.getLastDiv();
+            let iframeBody = $('#gutter');
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please define iframeBody as a variable in the attached() method for re-use throughout the .js file. It's repetitive to continue to define it locally whenever you need it.

File: src/jsGutter/js-gutter.js [hunk 8]
```diff
@@ -1,120 +1,152 @@
...
-   CreateLine(line){
-      let iframeBody = $('#gutter');  
-      let indexOfDiv = this.GetLastDiv();
-      for(indexOfDiv ; indexOfDiv <=line; indexOfDiv++){ 
-         iframeBody.append("<div id=line"+indexOfDiv+"></div>");
-         iframeBody.find("#line"+indexOfDiv).addClass("line_height");
-      }
-   }
-   GetLastDiv(){
         let iframeBody = $('#gutter');
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please define iframeBody as a variable in the attached() method for re-use throughout the .js file. It's repetitive to continue to define it locally whenever you need it.

File: src/jsGutter/js-gutter.js [hunk 9]
```diff
@@ -1,120 +1,152 @@
...
-
-     
-      
-      
-}
+        iframeBody.find("#line" + line).append(" [ " + contents + " ] ");
+    }
+
+    CreateLine(line) {
+        let iframeBody = $('#gutter');
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please define iframeBody as a variable in the attached() method for re-use throughout the .js file. It's repetitive to continue to define it locally whenever you need it.

File: src/jsGutter/js-gutter.js [hunk 10]
```diff
@@ -1,120 +1,152 @@
...
+        let indexOfDiv = this.getLastDiv();
+        for (indexOfDiv; indexOfDiv <= line; indexOfDiv++) {
+            iframeBody.append("<div id=line" + indexOfDiv + "></div>");
+            iframeBody.find("#line" + indexOfDiv).addClass("line_height");
+        }
+    }
+    getLastDiv() {
+        let iframeBody = $('#gutter');
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please define iframeBody as a variable in the attached() method for re-use throughout the .js file. It's repetitive to continue to define it locally whenever you need it.

File: src/jsGutter/js-gutter.js [hunk 11]
```diff
@@ -1,120 +1,152 @@
...
+        let indexOfDiv = 1;
+        while (iframeBody.find('#line' + indexOfDiv).length != 0) {
+            indexOfDiv++;
+        }
+        return indexOfDiv;
+    }
+    removeLine(lastline, lastDiv) {
+        let iframeBody = $('#gutter');
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please define iframeBody as a variable in the attached() method for re-use throughout the .js file. It's repetitive to continue to define it locally whenever you need it.

File: src/jsGutter/js-gutter.js [hunk 12]
```diff
@@ -1,120 +1,152 @@
...
+        while (lastline < lastDiv) {
+            iframeBody.find('#line' + lastDiv).remove();
+            lastDiv--;
+        }
+    }
+    highlightLine(line, lastline) {
+
+        let lastDiv = this.getLastDiv();
+        let iframeBody = $('#gutter');
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please define iframeBody as a variable in the attached() method for re-use throughout the .js file. It's repetitive to continue to define it locally whenever you need it.

File: src/jsGutter/js-gutter.js [hunk 13]
```diff
@@ -1,120 +1,152 @@
...
+        iframeBody.find("#line" + selectedLine).removeClass("highlight_gutter");
+        iframeBody.find("#line" + line).addClass("highlight_gutter");
+        this.selectedLine = line;
+
+
+
+    }
 
+    clearGutter() {
+        let iframeBody = $('#gutter');
```
Comment (jbhilton11) [inline src/jsGutter/js-gutter.js]: Please define iframeBody as a variable in the attached() method for re-use throughout the .js file. It's repetitive to continue to define it locally whenever you need it.

**Comment by Alaboudi1** [comment]
Thank you @jbhilton11 & @luminaxster for the code review!

File: src/jsEditor/js-editor.js [hunk 14]
```diff
@@ -50,33 +52,42 @@ export class JsEditor {
...
-         ea.publish('onJsEditorChanged', {
-            js: js, 
-            length: session.getLength(), 
+        let curs = editor.getCursorPosition().row + 1;
+
+        // This line strip out the spaces at the end of the documents.
+        let newStr = js.replace(/(\s+$)/g, '');
+        // then, hash it and store it in localHash variable.
+        let localHash = md5(newStr);
+        if (editorHashedText != localHash ) {
```
Comment (jbhilton11) [inline src/jsEditor/js-editor.js]: I would do !== instead of !=

File: src/jsGutter/js-gutter.js [hunk 15]
```diff
@@ -1,120 +1,147 @@
 /* global Firepad */
 /* global Firebase */
 /* global ace */
+import {
```
Comment (luminaxster) [inline src/jsGutter/js-gutter.js]: You don't need this import =P

File: src/jsEditor/js-editor.js (hunk 14 above)
Comment (Alaboudi1) [inline src/jsEditor/js-editor.js]: This is what I have tried the first time. Strangely, this did not work at that time. However, it works now as expected :|  . Maybe that is due to injecting the md5 library properly to the class? I do not know. Anyway, what matters is that it works fine now.

**Comment by jbhilton11** [comment]
I believe this looks fine with all of the fixes made. Merging into master.

**Pull request merged**